cdisc_rules_engine/services/datasetxpt_metadata_reader.py

```diff
@@ -18,40 +15,48 @@ class DatasetXPTMetadataReader:
     # TODO. Maybe in future it is worth having multiple constructors
     #  like from_bytes, from_file etc. But now there is no immediate need for that.
     def __init__(self, file_path: str, file_name: str):
-        with open(file_path, "rb") as f:
-            self._file_contents = f.read()
-        self._metadata_container = None
+        file_size = os.path.getsize(file_path)
+        if file_size > psutil.virtual_memory().available * 0.25:
+            self._estimate_dataset_length = True
+            self.row_limit = 1
+        else:
+            self._estimate_dataset_length = False
+            self.row_limit = 0
```

Comment:
This is one of the main causes of the large processing time. We are only asking for metadata and we only need `dataset` in order to get the `DOMAIN` from the first row. I think in all cases, we want the row_limit to be 1, otherwise we are reading in the entire dataset when we are retrieving the metadata.

Reply:
I don't think this was actually the cause because if we are using large datasets the row_limit should be 1. I kept the row_limit at 0 for smaller datasets because we want an exact record count in the metadata. For large datasets we are now getting the record count when we convert to parquet, so we don't need to read the whole dataset.

The reason it was reading the whole dataset even for large datasets is because of the issue you pointed out here:
https://github.com/cdisc-org/cdisc-rules-engine/pull/615#discussion_r1570990049